lib/table.c

```diff
@@ -2260,6 +2260,17 @@ simplifier_merge_ancestors(simplifier_t *self, node_id_t input_id)
         goto out;
     }
 
+    /* check if the input ID is a sample */
+    output_id = self->node_id_map[input_id];
+    coalescence = sample_node = self->is_sample[input_id];
```

Comment:
I don't understand this bit. Why is coalescence not always initially false? I'm missing something here I think, as if we can pull this outside the inner loop, then perhaps we should have separate simpler loops to take care of these cases. Perhaps we should deal with the case of ``is_sample`` separately?

Also, I'm not a big fan of multiple assignments (a = b = c). It's clear here what it means, but I have to think about it to make sure it's something more subtle.

Reply:
Here's the logic:

The `output_id` was previously named `v`, assigned either [here](https://github.com/jeromekelleher/msprime/blob/master/lib/table.c#L2307) or [here](https://github.com/jeromekelleher/msprime/blob/master/lib/table.c#L2330).  It could turn out that this ancestor has an output ID because:

- either we [encounter a coalescence](https://github.com/jeromekelleher/msprime/blob/master/lib/table.c#L2324) 
- or [discover that this input node is a sample](https://github.com/jeromekelleher/msprime/blob/master/lib/table.c#L2318) and so already has an output ID.

The logic above had us checking if `self->node_id_map[input_id]` existed several times in different places, but for essentially the same reason. 

Previously, we did:

- [this bit](https://github.com/jeromekelleher/msprime/blob/master/lib/table.c#L2308) only if the node in question was a sample.  This is the flag `sample_node`.
- [this bit](https://github.com/jeromekelleher/msprime/blob/master/lib/table.c#L2321) only if a coalescence has not been encountered yet **and** the node was [not a sample](https://github.com/jeromekelleher/msprime/blob/master/lib/table.c#L2321).  This is the flag `coalescence`.

The `coalescence` flag is assigned also if `sample_node` is because if it's a sample node, there will be edges output, and hence coalescence.

Reply:
I don't think coalescence flag is right, and could introduce subtle errors. Put it another way, what do we gain by not setting this to ``false``?

The other changes are great though, and definitely clarify. I'm happy to merge if we set ``coalescence = false`` rather than tying it to the sample status of the input node.